570 transfer logic documentation

andreievg opened the pull request:
closes #570 

These branch may not compile, check out [head PR here](https://github.com/openmsupply/open-msupply/pull/587) for full changes of transfer logic (if you are viewing/testing overall changes). PR changes will be done to that branch, with link to commit in comments they address.

I thought to commit documentation first, so it's present in all downstream PR's and would give a good introduction for requirements and implementation logic

clemens-msupply commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,42 @@
+# Changelog
+
```
Think an intro sentence would be good (or move last sentence to the top):

The changelog records database change for selected tables, i.e. if a row has been upserted or deleted.

andreievg replied:
Thanks: f98a4420fc3ec99d5c5bb13e1676d4f72a9c8717

clemens-msupply commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,42 @@
+# Changelog
+
+A number of operations rely on processing a `queue` of database changes. There include pushing records to central server during synchronisation, processing transfer records, processing messaging, aggregation, interoperability synchronisation, etc.. 
```
```suggestion
A number of operations rely on processing a `queue` of database changes.
These operations include pushing records to central server during synchronisation, processing transfer records, processing messaging, aggregation, interoperability synchronisation, etc.. 
```

To make reviewing easier I always put one sentence in one line when writing the docs. It makes it also much easier to find the sentence when you fix a reviewer comment.

clemens-msupply commented on server/service/src/processors/README.md:
```diff
@@ -0,0 +1,33 @@
+# Processors
+
```
Thats like a pub sub? one sender multiple subscribers? Think one more intro sentence what processors are and example of how they are used would be good.

andreievg replied:
Thanks: 5df2a2f517dbe511db23ba48b78ad99865d86573
Also: f91c7e1f53b76d819b19f25d5a10a28915a58c9c

clemens-msupply commented on server/service/src/processors/README.md:
```diff
@@ -0,0 +1,33 @@
+
+## TODO
+
+* Current triggers do not await for the message to be sent, and do not await for the operation to be completed, this is potentialy desirable for processors after sychronisation and can be implemented with a more sophisticated messaging system. 
```
Sorry for jumping ahead, do we need to consider server crashes? e.g. events in the channel get lost during a restart?

Edit: ok so the service just uses a trigger to let the processor start working the changelog? And when there is a crash the processor can continue where it stopped? sounds good :)
Think this file is more about `triggers`? rename?

andreievg replied:
> Think this file is more about triggers? rename?

Hmm, actual processor operations are defined in this folder, check out branch that's mentioned in the pr comment.

roxy-dao commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,42 @@
+
+A number of operations rely on processing a `queue` of database changes. There include pushing records to central server during synchronisation, processing transfer records, processing messaging, aggregation, interoperability synchronisation, etc.. 
+
+Generic mechanism for recording and querying these database changes was implemente in the form of `changelog` table.
```
```suggestion
Generic mechanism for recording and querying these database changes was implemented in the form of the `changelog` table.
```

roxy-dao commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,42 @@
+
+Database triggers insert new entries into changelog table with:
+
+* cursor <- sequencial number for order of database operation
```
```suggestion
* cursor <- sequential number for order of database operation
```

roxy-dao commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+`OutboundShipment` -> `InboundShipment`
+
+The term `transfer` arrises from the need to `transfer` these records from one `sync site` to another, in a scenario where pair of linked transfer records belong to two different `sites`.
```
```suggestion
The term `transfer` arises from the need to `transfer` these records from one `sync site` to another, in a scenario where a pair of linked transfer records belong to two different `sites`.
```

roxy-dao commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+Each processor (`shipment` and `requisition` transfer processor) keep track of changelog it has processed, and when it's triggerd it will try processing any records that have been changed since the last processed changelog entry.
+
+We **only** want to process shipments and requisitions that are `destined` for current site, thus changelogs are filtered by name_ids belonging to stores on that are active on current site.
```
```suggestion
We **only** want to process shipments and requisitions that are `destined` for current site, thus changelogs are filtered by name_ids belonging to stores that are active on current site.
```

roxy-dao commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+* Request Requisition (id: A) is created and set to `sent` status
+* Requsition Processor is triggered
+* Processor sees new requisition in change log and executes `CreateResponseRequisitionProcessor` which creates Response Requisition (id: B) and links it to Request Requisition (requisition_id: A)
+* Processor tries to process any sother change log (it re-quieries changelogs until result is empty), and sees newly created Response Requisition, and will executed `LikRequestRequisitionProcessor` and set's (requisition_id: B) for Request Requistion (id: A)
```
```suggestion
* Processor tries to process any other changelogs (it re-queries changelogs until result is empty), and sees the newly created Response Requisition, and will executed `LikRequestRequisitionProcessor` and sets (requisition_id: B) for Request Requistion (id: A)
```

roxy-dao commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Potential for Circular/Infinite processing
+
+Any event driven system has a potential for infinite looping. Processor conditions are carefully crafted to avoid these scenarior, but there is a possibility of an edge case that hasn't been considered. Care should be taken when adding changes and review transfer processor code.
```
```suggestion
Any event driven system has a potential for infinite looping. Processor conditions are carefully crafted to avoid these scenarios, but there is a possibility of an edge case that hasn't been considered. Care should be taken when adding changes and review transfer processor code.
```

roxy-dao commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Referential integrity
+
+Having referential constraints in database means that at access time we dont' need to do extra checks, but referential constrains cause a slightly problem, for example `invoice_line` would be transfered from source site to destination site, but associated `stock_line` and `location` will not (as they would only exists on source site). This is metigated by incoming sync translator, if a record doe ...
```
```suggestion
Having referential constraints in database means that at access time we don't need to do extra checks, but referential constrains cause a slight problem, for example `invoice_line` would be transferred from source site to destination site, but associated `stock_line` and `location` will not (as they would only exists on source site). This is mitigated by incoming sync translator, if a record does not belong to the current site both stock_line and location are set to null in database
```

roxy-dao commented on server/service/src/processors/transfer/shipment/README.md:
```diff
@@ -0,0 +1,19 @@
+
+## Invoice Line on Update
+
+When outbound shipment is updated and inbound shipment was already generated, invoice lines will be dropped and and reinstanted to match outbound shipment (this is the simples way to update potentially changed invoice lines)
```
```suggestion
When an outbound shipment is updated and an inbound shipment is already generated, invoice lines will be dropped and reinstated to match the outbound shipment (this is the simplest way to update potentially changed invoice lines)
```

jmbrunskill commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,43 @@
+
+For some data tables it's not possible to have foreign records on the site, thus name_id and store_id are optional and are set to null for those cases.
+
+## Cursors
```
If you are targeting understanding for someone completely new to the code base, I would be tempted to move the Cursors section up earlier in the document.
And I'd add a quick explanation something like this:

```
The 'cursor' is used by changelog consumers to track which records they have already processed and allow them to start processing again from where they left off.
For example, if the requisition processor has processed all the records until cursor 10, it should record the last cursor it processed (e.g. 10). Next time it starts processing records from the changelog, it will start from cursor number 11. 
Meanwhile the shipment processor could be further ahead, processing cursor number 50. By allowing the processors to track their own cursor value, we don't need a separate queue of unprocessed records for each processor.
 ```

andreievg replied:
Thanks: 5707429b6383efb1ddbd03b77b08316453bfc2fe

> If you are targeting understanding for someone completely new to the code base

Hmm, this is more of an advanced topic in my opinion, but yeah good to be more specific.

jmbrunskill commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,43 @@
+* cursor <- sequential number for order of database operation
+* record_id 
+* table_name
+* store_id
```
```suggestion
* store_id <- the store that `owns` the record
* name_id <- the name_id of the `destination` entity (often also a store)
```

Not sure if this needs to be here, but might help someone understand the distinction. I'm guessing there might be exceptions to these rules?

andreievg replied:
I've change `(often also a store)` to just `(can also be a store)` 5707429b6383efb1ddbd03b77b08316453bfc2fe

jmbrunskill reviewed (approved):
I still see a few spelling mistakes (such as `chanells` instead of `channels`) for example.
Would it be helpful if @roxy-dao or I did a check for these typos and just committed against this branch rather than people adding tons of comments about typos?

I've also made a couple of suggestions that you might want to include.

andreievg commented:
>I still see a few spelling mistakes (such as chanells instead of channels) for example.
Would it be helpful if @roxy-dao or I did a check for these typos and just committed against this branch rather than people adding tons of comments about typos?

Updated spelling mistake here: 5707429b6383efb1ddbd03b77b08316453bfc2fe

@roxy-dao @jmbrunskill, agree, it would be more efficient to do commits straight to the branch, or even better, what I usually do for suggested changes, make a branch from this branch with those commits and link in github `compare` link (author can do the merge)

> I've also made a couple of suggestions that you might want to include.

included, thanks

mark-prins commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,46 @@
+A number of operations rely on processing a `queue` of database changes.
+These operations include pushing records to central server during synchronisation, processing transfer records, processing messaging, aggregation, interoperability synchronisation, etc.. 
+
+Generic mechanism for recording and querying these database changes was implemented in the form of the `changelog` table.
```
:nit
```suggestion
A generic mechanism for recording and querying these database changes was implemented in the form of the `changelog` table.
```

mark-prins commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## Triggers
+
+Database triggers insert new entries into changelog table with:
```
:nit
```suggestion
Database triggers insert new entries into the `changelog` table with:
```

though I'm also accustomed, thanks to mSupply, to use the syntax `[tablename]` when referring to db tables and like that now

andreievg replied:
Will try to use [] that in the future, thanks

mark-prins commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,46 @@
+* record_id 
+* table_name
+* store_id <- the store that `owns` the record
+* name_id <- the name_id of the `destination` entity (can also be a store)
```
side thought: would it be more helpful if the field was `destination_name_id`? it would therefore be self-documenting.

andreievg replied:
hmm, i like this but only if all other instances have the same, if it's name_id in requisition and invoice but it's different in changelog i think it becomes inconsistent

mark-prins commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## Cursors
+
+The 'cursor' is used by changelog consumers to track which records they have already processed and allow them to start processing again from where they left off.
```
the 'quotes' aren't required - it is a cursor that we're talking about

andreievg replied:
Thanks, it's from a suggestion btw: https://github.com/openmsupply/open-msupply/pull/571#discussion_r967939795, removed 256869b3c4be2e510ec52b029561f7f25d32771f

mark-prins commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,46 @@
+For example, if the requisition processor has processed all the records until cursor 10, it should record the last cursor it processed (e.g. 10). Next time it starts processing records from the changelog, it will start from cursor number 11. 
+Meanwhile the shipment processor could be further ahead, processing cursor number 50. By allowing the processors to track their own cursor value, we don't need a separate queue of unprocessed records for each processor.
+
+Services that use changelog need to manually maintain cursor and persist it using key_value_store repository (get_i64 and set_i64). After a query to changelog repository, next cursor would be `last cursor` in the query output `+ 1`.
```
:nit
```suggestion
Services that use the changelog need to manually maintain cursor and persist it using the `key_value_store` repository (get_i64 and set_i64). After a query to changelog repository, next cursor would be `last cursor` in the query output `+ 1`.
```

mark-prins commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,46 @@
+
+## name_id and store_id
+
+Some consumers of changelog need to filter database operations based on `ownership` of the record on current site. 
```
unsure what the backticks around _ownership_ are supposed to represent. I think that you intend actual ownership in which case they aren't required.

andreievg replied:
I need a guide on quotes, I thought that backticks are just there to emphasise something ? What is the conventions for backticks in documentation ?

mark-prins commented on server/repository/src/db_diesel/changelog/README.md:
```diff
@@ -0,0 +1,46 @@
+* Synchroniser should only push records belonging to current site
+* Transfer should only process records that are destined for current site
+
+name_id and store_id is stored against records that need to be filtered in this fashion against database operation in changelog. Extra logic is added to triggers for these records to transfer store_id and name_id to changelog.
```
unsure what the convention for database tables and columns is: elsewhere the table and column names are in \`backticks\`  but here we have store_id and name_id mentioned and they appear to be database columns, because they are in the xxx_id format, but they are in plain text. If they are referring to variables or values, it would be clearer if they were name ID and store ID ( or Id? )

andreievg replied:
Hmm, once again not sure about backticks, I think since the topic is about name_id and store_id I probably left them without emphasis

mark-prins commented on server/service/src/processors/README.md:
```diff
@@ -0,0 +1,33 @@
+
+## Implementation
+
+Channels are started via `Processors::init()`, and service provider takes the `Sender` part of the `Processors::init()` result tuple (`Sender` is already `Arc`ed). `Processors::spawn()` will start a task and return the `handle`. The task will wait on a message sent to trigger a particular processor and will invoke that processor.
```
:nit
```suggestion
Channels are started via `Processors::init()`, and service provider takes the `Sender` part of the `Processors::init()` result tuple (`Sender` is already `Arc'ed`). `Processors::spawn()` will start a task and return the `handle`. The task will wait on a message sent to trigger a particular processor and will invoke that processor.
```

andreievg replied:
Since the data type is `Arc` i wanted to keep the emphasis on the `Arc`

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+# Transfers
+
+Pair records that belong to two different stores but are intrinsically linked together are called `transfers`:
```
```suggestion
Paired records that belong to two different stores but are intrinsically linked together are called `transfers`:
```

or 'Pairs of records...'

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+`OutboundShipment` -> `InboundShipment`
+
+The term `transfer` arises from the need to `transfer` these records from one `sync site` to another, in a scenario where a pair of linked transfer records belong to two different `sites`.
```
the initial definition doesn't make the distinction about being on a different site. If that's key to being a 'transfer' then it should be included too

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+![omSupply transfer workflow](./doc/omSupply_transfer_workflow.png)
+
+Big difference between mSupply transfer requirements and omSupply:
+* The inbound shipment part of the transfer is generate before outbound shipment is shipped (before inbound shipment is editable)
+* Status updates on destinaton part of transfer (response requisition and inbound shipment) are reflected on source part of the transfer (request requisition and outbound shipment)
```
```suggestion
The big difference between mSupply transfer requirements and omSupply:
* The inbound shipment part of the transfer is generated before the outbound shipment is shipped (before the inbound shipment is editable)
* Status updates on the destination part of transfer (response requisition and inbound shipment) are reflected on the source part of the transfer (request requisition and outbound shipment)
```

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+## Ownership
+
+In our synchronisation system, we have a strong rule of:
+* records can belong to **one** site only, which mean they can only be mutated on that site.
```
```suggestion
* records can belong to **one** site only, which means they can only be mutated on that site.
```

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+In our synchronisation system, we have a strong rule of:
+* records can belong to **one** site only, which mean they can only be mutated on that site.
+
+This means the `other half` of transfer record needs to be generated on the site that will own that record. This is done by processing `changelog` after syncrhonisation. 
```
```suggestion
This means the corresponding half of the transfer record needs to be generated on the site that will own that record. This is done by processing the `changelog` after synchronisation. 
```

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Changelog Processing
+
+Each processor (`shipment` and `requisition` transfer processor) keep track of changelog it has processed, and when it's triggerd it will try processing any records that have been changed since the last processed changelog entry.
```
```suggestion
Each processor (`shipment` and `requisition` transfer processor) keeps track of the changelog records that it has processed, and when triggered it will try processing any records that have been changed since the last processed changelog entry.
```

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+Each processor (`shipment` and `requisition` transfer processor) keep track of changelog it has processed, and when it's triggerd it will try processing any records that have been changed since the last processed changelog entry.
+
+We **only** want to process shipments and requisitions that are `destined` for current site, thus changelogs are filtered by name_ids belonging to stores that are active on current site.
```
```suggestion
We **only** want to process shipments and requisitions that are destined for the current site, thus changelogs are filtered by `name_id`s belonging to stores that are active on the current site.
```

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+We **only** want to process shipments and requisitions that are `destined` for current site, thus changelogs are filtered by name_ids belonging to stores that are active on current site.
+
+## If both stores are active on current site
```
```suggestion
## If both stores are active on the current site
```

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## If both stores are active on current site
+
+Same exact process is used, except the processors are triggered in services, see diagram below. This diagram doesn't quite show how records are `linked` using processors, explained in this example
```
```suggestion
Same exact process is used, except the processors are triggered in services, see diagram below. This diagram doesn't quite show how records are linked using processors, explained in this example
```

andreievg replied:
I think the emphasis should be on `linked` here

mark-prins commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Referential integrity
+
+Having referential constraints in database means that at access time we don't need to do extra checks, but referential constrains cause a slight problem, for example `invoice_line` would be transferred from source site to destination site, but associated `stock_line` and `location` will not (as they would only exists on source site). This is mitigated by incoming sync translator, if a record does ...
```
```suggestion
Having referential constraints in the database means that at access time we don't need to do extra checks, but referential constraints cause a slight problem. For example `invoice_line` would be transferred from source site to destination site, but associated `stock_line` and `location` will not (as they would only exist on the source site). This is mitigated by the incoming sync translator, if a record does not belong to the current site both `stock_line` and `location` are set to null in the database
```

mark-prins commented on server/service/src/processors/transfer/shipment/README.md:
```diff
@@ -0,0 +1,19 @@
+
+## Request Requisition linked to Inbound Shipment
+
+When inbound shipment is created from outbound shipment, we check for requisition linked to outbound shipment (response requisition) and check what request requisition is linked to that response requisition and link inbound shipment with correct request requisition
```
```suggestion
When an inbound shipment is created from outbound shipment, we check for a requisition linked to the outbound shipment (response or client requisition) and check which request requisition is linked to that response requisition and link the inbound shipment with correct request requisition
```

andreievg replied:
Is `client requisition` a new type ? I think it's called `customer requisition` ? I was hoping to avoid using legacy terms in omSupply, added suggestion as a commit, without `or client`: 6225fa298f3bc269f623b520aed754ae1d7a6af5

mark-prins replied:
sorry, yeah. I meant customer 🙄

mark-prins commented on server/service/src/sync/test/integration/README.MD:
```diff
@@ -91,13 +92,27 @@ For each step:
+
+Using `RequisitionTransferTester` and `ShipmentTransferTester` defined in transfer processors unit test.
+
+There structs implement test methods that need to be run sequentially, they create, update, delete source and destination records and test that corresponding transfer records have been changed accordingly. Each of those methods accept connection or service provider to allow the operation to be executed on a chosen site.
```
```suggestion
These structs implement test methods that need to be run sequentially. They create, update, delete source and destination records and test that corresponding transfer records have been changed accordingly. Each of those methods accept connection or service provider to allow the operation to be executed on a chosen site.
```

mark-prins reviewed (commented):
🎉  brilliant! great to have such comprehensive docs.

I've picked nits mostly.. I think there are useful suggestions in there too

andreievg commented:
Thank you @roxy-dao @jmbrunskill @clemens-msupply @mark-prins for your reviews ! I think I've addressed all of the comments, will keep an eye on any further comments before merging of head PR

roxy-dao commented on server/service/src/processors/README.md:
```diff
@@ -0,0 +1,43 @@
+
+## Motivation
+
+It is desirable to perform some operations syncrhonously as they may cause undesirable and unexpected side effects when run concurrently. These operations should be `triggerable` in services, wherever `ServiceContext` and `ServiceProvider` are available and should be `triggerable` on individual basis.
```
```suggestion
It is desirable to perform some operations synchronously as they may cause undesirable and unexpected side effects when run concurrently. These operations should be `triggerable` in services, wherever `ServiceContext` and `ServiceProvider` are available and should be `triggerable` on individual basis.
```

roxy-dao commented on server/service/src/processors/transfer/README.md:
```diff
@@ -0,0 +1,62 @@
+
+Same exact process is used, except the processors are triggered in services, see diagram below. This diagram doesn't quite show how records are `linked` using processors, explained in this example
+* Request Requisition (id: A) is created and set to `sent` status
+* Requsition Processor is triggered
```
```suggestion
* Requisition Processor is triggered
```

😄

andreievg replied:
🙏

andreievg commented:
Merging into: https://github.com/openmsupply/open-msupply/pull/587
